### PR TITLE
chore(profiling): move profiler tests to riot

### PR DESCRIPTION
**Pull request by mabdinur**
## Description

Moves profiler tests from tox to riot.

Changes to gevent versions used in tests:
 - Pins greenlet to v0.4.16 for `gevent<20.9.0` (in python>=3.7) to resolve greenlet-gevent incompatibilities. 
 - Runs tests with the minimum required version of gevent and greenlet that is supported across all platforms. Avoids installing specific gevent versions for windows 32bit machines. 
 - Avoids installing gevent with --no-binary option. This flag is was [deprecated in pip v22.3](https://pip.pypa.io/en/stable/news/#deprecations-and-removals).
 - Runs profiler tests with the following scenarios for `python>=2.7,<=3.10`:
   - profiler with the **latest** supported versions of tenacity, gevent and protobuf
   - profiler with the **minimum** required versions of tenacity, gevent, and protobuf
   - profiler with the **latest** supported versions of tenacity, and protobuf (No gevent monkey patching)
   - ~~profiler with the **minimum** required versions for tenacity, and protobuf (No gevent monkey patching)~~  (this scenario was removed, it is a bit redundant)

## Remaining Work

python2.7 uwsgi tests are NOT compatible with riot venvs and are skipped in this PR:
- Running uwsgi app in a subprocess raises the following error:  [example](https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/24236/workflows/17d8c6b1-2835-47cc-8548-8ac2f47dd720/jobs/1647017)
- Note: installing legacy versions of gevent fail on M1 laptops. Use the dd-tracy-py [docker image](https://hub.docker.com/r/datadog/dd-trace-py) to run the profiling test suite. 
  - Error from local test run: 
```
 *** Operational MODE: single process ***
Traceback (most recent call last):
  File "/Users/munirabdinur/go/src/github.com/DataDog/dd-trace-py/tests/profiling/uwsgi-app.py", line 1, in <module>
    import ddtrace.profiling.auto  # noqa
  File "./ddtrace/internal/module.py", line 182, in _load_module
    module = self.loader.load_module(fullname)
  File "/Users/munirabdinur/.pyenv/versions/2.7.18/lib/python2.7/pkgutil.py", line 243, in load_module
    mod = imp.load_module(fullname, self.file, self.filename, self.etc)
  File "/Users/munirabdinur/go/src/github.com/DataDog/dd-trace-py/ddtrace/profiling/__init__.py", line 5, in <module>
    from .profiler import Profiler  # noqa:F401
  File "./ddtrace/internal/module.py", line 182, in _load_module
    module = self.loader.load_module(fullname)
  File "/Users/munirabdinur/.pyenv/versions/2.7.18/lib/python2.7/pkgutil.py", line 243, in load_module
    mod = imp.load_module(fullname, self.file, self.filename, self.etc)
  File "/Users/munirabdinur/go/src/github.com/DataDog/dd-trace-py/ddtrace/profiling/profiler.py", line 29, in <module>
    from ddtrace.profiling.exporter import file
  File "./ddtrace/internal/module.py", line 182, in _load_module
    module = self.loader.load_module(fullname)
  File "/Users/munirabdinur/.pyenv/versions/2.7.18/lib/python2.7/pkgutil.py", line 243, in load_module
    mod = imp.load_module(fullname, self.file, self.filename, self.etc)
  File "/Users/munirabdinur/go/src/github.com/DataDog/dd-trace-py/ddtrace/profiling/exporter/file.py", line 7, in <module>
    from ddtrace.profiling.exporter import pprof
  File "./ddtrace/internal/module.py", line 182, in _load_module
    module = self.loader.load_module(fullname)
  File "/Users/munirabdinur/.pyenv/versions/2.7.18/lib/python2.7/pkgutil.py", line 243, in load_module
    mod = imp.load_module(fullname, self.file, self.filename, self.etc)
  File "ddtrace/profiling/exporter/pprof.pyx", line 96, in init ddtrace.profiling.exporter.pprof
    _pb_version = _protobuf_version()
  File "ddtrace/profiling/exporter/pprof.pyx", line 89, in ddtrace.profiling.exporter.pprof._protobuf_version
    import google.protobuf
ImportError: No module named google.protobuf
unable to load app 0 (mountpoint='') (callable not found or import error)
*** no app loaded. GAME OVER ***
``` 

## Reviewer Checklist
- [ ] Title is accurate.
- [ ] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -512,18 +504,18 @@ jobs:
...
       # circleci/windows@5.0 orb includes python 3.10.6
-      - run_tox_scenario:
+      - run_test:
           store_coverage: false
-          pattern: '^py310-profile'
+          pattern: 'profiler'
 
   profile:
     <<: *contrib_job
+    parallelism: 7
```
Comment (mabdinur) [inline .circleci/config.yml:513]:
```suggestion
    parallelism: 15
```

**Comment by P403n1x87** [comment]
> python2.7 uwsgi tests are NOT compatible with riot venvs and are skipped in this PR


@mabdinur riot can create a "fat" venv with `create=True` e.g.
~~~python
Venv(
    name="foo",
    cmd="pytest ...",
    pkgs={...},
    pys="2.7",
    create=True,
)
~~~
This will create a single venv with all the dependencies installed into it. Can we give this a try to see if it fixes the problem with 2.7?

**Review by P403n1x87 — DISMISSED** [review]

**Comment by mabdinur** [comment]
Blocked by: https://github.com/DataDog/riot/pull/199

**Comment by Yun-Kim** [comment]
I believe the riot changes are missing the profiler tests for Python 3.11. We can just add a new Venv as such:
```python
Venv(
    pys=select_pys(min_version="3.11"),
    pkgs={"pytest-asyncio": latest},
    command="python -m tests.profiling.run pytest --no-cov --capture=no --verbosity=2"
    " --benchmark-disable {cmdargs} tests/profiling",
    venvs=[
        # profiler with minreqs
        Venv(
            pkgs={
                "gunicorn": latest,
                "tenacity": "==7.0.0",
                "protobuf": "~=3.0",
                "gevent": "==22.10.2",
                "greenlet": "==2.0.1",
            },
            env={"DD_PROFILE_TEST_GEVENT": "1"},
        ),
        # profiler with latest reqs
        Venv(
            pkgs={
                "gunicorn[gevent]": latest,
                "protobuf": latest,
                "tenacity": latest,
            },  
            env={"DD_PROFILE_TEST_GEVENT": "1"},
        ), 
        # profiler without gevent
        Venv(
            pkgs={
                "gunicorn": latest,
                "protobuf": latest,
                "tenacity": latest,
            },
        ),
    ],
),
```

**Comment by Yun-Kim** [comment]
It looks like creating the "fat" venvs in Python 2.7 are causing import problems (packages we have in ddtrace's setup.py `install_requires` like jsonschema, etc. are not present in the Python 2.7 venvs).

File: riotfile.py [hunk 2]
```diff
@@ -293,6 +293,165 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
...
+                    pkgs={"pytest-asyncio": latest},
+                    command="python -m tests.profiling.run pytest --no-cov --capture=no --verbosity=2"
+                    " --benchmark-disable {cmdargs} tests/profiling",
+                    venvs=[
+                        # profiler with minreqs
+                        # https://github.com/gevent/gevent/issues/1674
+                        Venv(
+                            pkgs={
+                                "gunicorn": latest,
+                                "tenacity": "==5.0.1",
```
Comment (Yun-Kim) [inline riotfile.py:427]: Python 3.11 requires `tenacity>=7.0.0`

**Comment by emmettbutler** [comment]
I'm having a bit of a hard time understanding from this PR's discussion whether it's ready for review or if there are additional changes planned. @mabdinur would you like a review on this in its current state?

**Comment by emmettbutler** [comment]
This has been superseded by #5633

**Pull request closed without merging**